lib/puppet/type/file.rb

```diff
@@ -220,6 +220,22 @@ def self.title_patterns
     end
   end
 
+  newparam(:max_files) do
+    desc "In case the resurce is a directory and the recursion is enabled, puppet will
+      generate a new resource for each file file found, possible leading to
+      an excessive number of resources generated without any control.
+
+      Setting `max_files` will check the number of file resources that
+      will eventually be created and will raise a resource argument error if the
+      limit will be exceeded.
+
+      Use value `0` to disable the check. In this case, a warning is logged if
+      a soft limit (hardcoded to 1000) is exceeded."
```

Comment:
nit "soft limit (hardcoded ...)" is a little confusing. Maybe drop soft/hard and just say "...a warning is logged if the number of files exceeds 1000."?